system/include/emscripten/val.h

```diff
@@ -385,116 +385,123 @@ class val {
   explicit val(T&& value) {
     using namespace internal;
 
-    typedef internal::BindingType<T> BT;
     WireTypePack<T> argv(std::forward<T>(value));
-    handle = _emval_take_value(internal::TypeID<T>::get(), argv);
+    new (this) val(_emval_take_value(internal::TypeID<T>::get(), argv));
   }
 
-  val() : handle(EM_VAL(internal::_EMVAL_UNDEFINED)) {}
+  val() : val(EM_VAL(internal::_EMVAL_UNDEFINED)) {}
 
   explicit val(const char* v)
-      : handle(internal::_emval_new_cstring(v))
+      : val(internal::_emval_new_cstring(v))
   {}
 
-  val(val&& v) : handle(v.handle) {
-    v.handle = 0;
+  val(val&& v) : handle_(v.handle_)
+#if !defined(NDEBUG) && defined(_REENTRANT)
+      , thread(v.thread)
+#endif
+  {
+    v.handle_ = 0;
   }
 
-  val(const val& v) : handle(v.handle) {
-    internal::_emval_incref(handle);
+  val(const val& v) : handle_(v.handle_)
+#if !defined(NDEBUG) && defined(_REENTRANT)
+      , thread(v.thread)
+#endif
+  {
+    internal::_emval_incref(handle_);
```

Comment:
Should this use `as_handle` (same with the destructor below)?

Reply:
Yeah just noticed those too and updating. Thanks!

Reply:
Pushed.